server/neptune/workflows/internal/pr/runner.go

```diff
@@ -0,0 +1,217 @@
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/terraform/state"
+	"go.temporal.io/sdk/temporal"
+	"go.temporal.io/sdk/workflow"
+)
+
+type TFWorkflow func(ctx workflow.Context, request terraform.Request) (terraform.Response, error)
+
+type Action int64
+
+const (
+	onNewRevision Action = iota
+	onShutdown
+)
+
+type RunnerState int64
+
+const (
+	working RunnerState = iota
+	waiting
+	complete
+)
+
+type Runner struct {
+	RevisionSignalChannel workflow.ReceiveChannel
+	RevisionReceiver      *receiver.RevisionReceiver
+	ShutdownSignalChannel workflow.ReceiveChannel
+	ShutdownReceiver      *receiver.ShutdownReceiver
+	Scope                 workflowMetrics.Scope
+	TFWorkflow            TFWorkflow
+	TFStateReceiver       StateReceiver
+
+	// mutable state
+	state                 RunnerState
+	lastAttemptedRevision string
+}
+
+func newRunner(ctx workflow.Context, scope workflowMetrics.Scope, tfWorkflow TFWorkflow, internalNotifiers []WorkflowNotifier) *Runner {
+	revisionReceiver := receiver.NewRevisionReceiver(ctx, scope)
+	shutdownReceiver := receiver.NewShutdownReceiver(ctx, scope)
+	return &Runner{
+		RevisionSignalChannel: workflow.GetSignalChannel(ctx, receiver.TerraformRevisionSignalID),
+		RevisionReceiver:      &revisionReceiver,
+		ShutdownSignalChannel: workflow.GetSignalChannel(ctx, receiver.ShutdownSignalID),
+		ShutdownReceiver:      &shutdownReceiver,
+		Scope:                 scope,
+		TFWorkflow:            tfWorkflow,
+		TFStateReceiver:       StateReceiver{InternalNotifiers: internalNotifiers},
+	}
+}
+
+// Run handles managing the workflow's context lifecycles as new signals/poll events are received and
+// change the current PRAction status
+func (r *Runner) Run(ctx workflow.Context) error {
+	var action Action
+	var prRevision receiver.Revision
+
+	//TODO: add approve signal, timeouts, poll variation of shutdown signal
+	s := workflow.NewSelector(ctx)
+	s.AddReceive(r.RevisionSignalChannel, func(c workflow.ReceiveChannel, more bool) {
+		prRevision = r.RevisionReceiver.Receive(c, more)
+		action = onNewRevision
+	})
+
+	s.AddReceive(r.ShutdownSignalChannel, func(c workflow.ReceiveChannel, more bool) {
+		r.ShutdownReceiver.Receive(c, more)
+		action = onShutdown
+		r.state = complete
+	})
+
+	_, cancel := workflow.WithCancel(ctx)
+	for {
+		s.Select(ctx)
+		switch action {
+		case onNewRevision:
+			revisionCtx := workflow.WithValue(ctx, internalContext.SHAKey, prRevision.Revision)
+			workflow.GetLogger(revisionCtx).Info("received revision")
+			if process := r.shouldProcessRevision(prRevision); !process {
+				continue
+			}
+			// cancel in progress workflow (if it exists) and start up a new one
+			cancel()
+			revisionCtx, cancel = workflow.WithCancel(revisionCtx)
+			r.state = working
+			r.lastAttemptedRevision = prRevision.Revision
+			workflow.Go(revisionCtx, func(c workflow.Context) {
+				r.processRevision(c, prRevision)
+			})
```

Comment:
if you move this to a separate function than u can just reuse the name `ctx`.
it's not standard practice to rename ctx, that makes this error prone since i could go in and just use the wrong context, I think i mentioned this before in our sync.